### PR TITLE
AI Problem Solver (frontend and backend linked)

### PR DESCRIPTION
**Pull request by hjamaly**
**Summary**
This PR introduces significant enhancements to the messaging system, including support for image-based messages and AI-powered suggestions, alongside various performance and UX improvements. The key updates are:

**Image-Based Messaging:**

- Refactored the backend integration to support image attachments.
- Implemented the ability to choose an image from the gallery and attach it to the message send bar.
- Fully integrated backend and frontend functionality for seamless image-based messaging.

**AI Solver Suggestions:**

- Added AI logic to analyze uploaded problem images:
- Determines if the user can solve their problem directly.
- Advises the user to create a request if further assistance is needed.

**UX Improvements:**

- Added a progress indicator when fetching data from Firebase to enhance user experience by providing real-time feedback.

**Testing:**

- Added unit tests for all new methods in the ChatRepository (Firestore).
- Ensured the backend methods are fully tested.
- Tested frontend components for new functionalities.
- Achieved ~65% line coverage for the new code, ensuring it does not lower the overall project coverage, which remains at 80%+.

Note: Some sections, such as simulating AI-generated responses, are challenging to test directly. These will be addressed with end-to-end (E2E) tests using a local emulator in the next sprint.

**Future Work:**

Implement E2E tests with a local emulator to cover currently hard-to-test parts of the code, ensuring complete test coverage and robustness of the new functionalities.
[Screen_recording_20241205_204539.webm](https://github.com/user-attachments/assets/cc10c802-343b-474d-8b5b-3555ab650ffc)

### CODE REVIEW

File: app/src/main/java/com/android/solvit/shared/model/utils/ImageLoader.kt
Comment (yshdb5) [inline app/src/main/java/com/android/solvit/shared/model/utils/ImageLoader.kt:40]: You should clean up commented code

File: app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt
Comment (yshdb5) [inline app/src/main/java/com/android/solvit/shared/ui/chat/ChatScreen.kt:110]: Just go over all the modified files to delete or uncomment commented code.

**Review by yshdb5 — CHANGES_REQUESTED** [review]
### Overall Comment
Great job enhancing the chat functionality! Integrating the new `AiSolverViewModel` and supporting AI-based interactions significantly improve the user experience by offering more dynamic chat options, including handling text and image messages together.

### Strong Elements
- **Integration of `AiSolverViewModel`**: Introducing `AiSolverViewModel` is well-executed, making the AI solver functionality modular and easier to manage.
- **Extended `ChatMessage` Class**: Extending `ChatMessage` to include `TextImageMessage` allows for more flexible message types.
- **Enhanced Repository Functions**: Updating `ChatRepository` and `ChatRepositoryFirestore` to handle AI-related chat actions improves overall functionality.
- **UI Improvements**: Updating UI components to support new message types and AI interactions ensures a cohesive user experience. The loading indicator in `MessageBox` is a nice touch.
- **Code Consistency and Readability**: The code is consistent and readable, with good use of comments and naming conventions.

### Areas for Improvement
- **Redundant Code**: Consider removing commented-out sections in `ImageLoader.kt` to keep the codebase clean.
- **Unit Tests**: I know that you are currently implementing complementary tests for the viewModels, it will be good to push them before merging.
- **AI chat**: For now, every message coming from the AI model starts with "Hey Hassan". This would be adapted to the seeker using the chat and consider dropping the "Hey ..." after the first message. Another suggestion: add a loading animation while waiting for the response of the AI model.
- **Regular chat**: When clicking on the mail icon to access the inbox screen I get an infinite loading animation.

### Conclusion
Overall, this PR significantly enhances the app's chat features by integrating AI capabilities and extending message types. The code is well-structured and thoughtfully implemented. With minor ameliorations, this update will greatly improve the app's chat and AI functionality. Great work!

**Review by manunam2002 — APPROVED** [review]
### Feedback

This PR is a very important addition to the projet as it makes sending images in the chat robust, and finalises the AI problem solver. This is a much more complex task than it seems and you executed it perfectly, great job! 

### Areas For improvements
- Documentation: hard to understand parts of the code could be documented clearer.
- Behaviour with AI: I would say that for the chat assistant it is not necessary, but for the solver it could be interesting to find a way to provide him the images, we should look at this option in the next meeting

**Review by yshdb5 — APPROVED** [review]

**Pull request merged**
